### PR TITLE
fix(device-link): 断链后恢复远程控制链路

### LINKED ISSUE #1324: [Bug] 撤销授权后 device-link 远程控制永久失效：[DEVICE_LINK_NOT_CONNECTED] control link is closed
**提交人**: YeGGo
**客户端版本**: 0.1.25

---

## 现象

另一台机器无法再远程控制本机的项目，报错：

```
Error invoking remote method 'device-link:invoke': Error: [DEVICE_LINK_NOT_CONNECTED] control link is closed
```

撤销授权后即出现此问题，之后**重新授权、重新登录都无法恢复**远程控制。

## 复现步骤

1. 本机（被控端）在 Cindy 的授权管理中**撤销**了对某台机器的授权。
2. 回到另一台机器（主控端），尝试远程控制本机的项目。
3. 报错 `[DEVICE_LINK_NOT_CONNECTED] control link is closed`。
4. 尝试重新授权、重新登录（两边都试过），问题依旧，远程控制始终无法恢复。

## 期望行为

1. **恢复远程控制**：重新授权后，另一台机器应能重新远程控制本机项目。
2. **明确的状态提示**：即使暂时无法恢复，应用也应显示清晰的授权/连接状态（而非停留在永久静默失败），并给出可操作的修复指引。

## 补充

- 撤销授权发生在**被控端本机**。
- 环境：两台机器均为 Windows。
- 客户端版本、GitHub 作者等环境信息由系统自动附加，如有需要请在 issue 中补充更多设备信息。
---
**版本区域**: CN
**OS**: win32 x64 (10.0.19045)
**界面语言**: zh-CN

### PR DESCRIPTION
**Pull request by dashhuang**
### 摘要

修复显式关闭 control link 后，Desktop 仍把后续远程调用永久留在 `LINK_NOT_OPEN` 状态的问题。根因是关闭状态只在传输层生效，但 Desktop 没有在下一次业务调用前重新建立 link-open / link-accept 生命周期。

### 变更类型

- [x] `fix` 缺陷修复
- [ ] `refactor` / `perf` 重构或性能优化
- [x] `docs` / `test` / `chore` 文档、测试或工程维护（补充回归测试）

### 范围

- 关联 Issue / 需求：Fixes #1324
- 本 PR 包含：
  - 发送前遇到 `LINK_NOT_OPEN` 时，通过已有并发去重的 `openRemoteLink()` 重开链路，并且只重试一次。
  - 重开等待期间重新检查本地禁用设备列表，保持撤权后的 fail-closed 语义。
  - 具体 session topic 的订阅在 link 未 ready 时先建立 streaming link。
  - 显式 close 后，legacy listing 的 `invoke-result` 仅按精确 request ID 放行一次，未知或重复 ID 仍拒绝。
- 明确不包含：
  - 不重试 `NOT_CONNECTED` / `relay connection lost`、timeout、backpressure 或已进入 `inFlight` 的请求；这些请求可能已经到达 relay/目标端，盲目重试会造成远程写操作重复执行。
  - 不修改 relay 服务端或 wire payload。
- 用户可见变化：重新授权或重新打开远程控制后，普通远程调用可以自动恢复；无新增 UI。
- 是否存在 breaking change：无

### UI 变化

- 不涉及：仅修改 Desktop main、共享 device-link client 及其单测，没有界面、交互或文案变化。
- 引用的设计规范：不涉及 UI。

### 自动验证

```text
pnpm --filter @cindy/device-link build
结果：通过

pnpm --filter @cindy/device-link test
结果：6 个测试文件、143 个测试通过

pnpm --filter desktop exec vitest run src/main/device-link/__tests__/linkRecovery.test.ts
结果：7 个测试通过

NODE_OPTIONS='--max-old-space-size=8192' pnpm --filter desktop typecheck
结果：通过

pnpm check:dco
结果：通过（1 个 commit，含 Signed-off-by）

bash /Users/dash/Code/XD/dash/Skills/git/scripts/run-unit-gate.sh /Users/dash/Code/Cindy/cindy-fix-device-link-reopen
结果：278/279 个测试文件、2998/2999 个测试通过；唯一失败为 apps/mobile/src/__tests__/startupSplashOverlay.test.ts 的 Expo introspect 子进程超过 5 秒超时。
```

### 手工验证

不涉及：本次未进行两台真实设备的远程控制手测。

### 未执行的验证

- 未执行真实 relay 断线/重连场景的手测；本 PR 对 `relay connection lost`（已 in-flight）保持不重试。
- 上述 mobile splash 测试失败已在未修改 mobile 文件的基线 checkout（HEAD `267f0a001`）单独复现，未将无关 mobile 修复混入本 PR。

### 风险分类

- [x] 协议兼容
- [x] 跨平台差异
- [ ] SQLite / migration
- [ ] system prompt
- [ ] 权限 / 安全 / 用户数据
- [ ] 存量插件兼容（批准状态 / 指纹 / manifest 校验 / 安装布局 / 包格式）

### 影响与回滚

- 影响范围：Desktop device-link 控制端调用路径与共享客户端的 closed-link legacy 回程；没有新增 wire 字段，旧端仍可按原 legacy/reliable 能力工作。已知边界是 `remoteSubscribe` 在重开完成后到实际发送之间再次断链时仍可能返回 `LINK_NOT_OPEN`，作为后续收敛项，不改变本 PR 的单次重试安全边界。
- 回滚 / 降级方式：回滚本 PR 即可恢复原行为；无需数据迁移。发生 relay 断线时仍沿用原有 `NOT_CONNECTED` / `inFlight` 错误语义，避免重复执行远程写操作。

### 提交前检查

- [x] 已 review 完整 diff
- [x] 每个 commit 都带 DCO 签名（`git commit -s`，见 [DCO](../DCO)）
- [x] UI 改动已在「UI 变化」注明引用的设计规范章节（不涉及 UI 则跳过）
- [x] 未提交凭证、令牌或授权文件
- [x] 已补充必要文档
- [x] 已确认测试结果或说明未执行原因

### CODE REVIEW

**Comment by MagicLizi** [comment]
@dashhuang 👋 这个 PR 还有 2 条 review conversation 没 resolve(apps/desktop/src/main/device-link/index.ts),auto-review 因此暂时跳过、没法继续审查 / 合并。

如果你已经按评论改完或回应了,请到对应 thread 上点 **Resolve conversation**;全部 resolve 后,下一轮 auto-review 会自动重新审查这个 PR。

File: apps/desktop/src/main/device-link/index.ts [hunk 1]
```diff
@@ -918,8 +930,16 @@ export async function remoteInvoke(
...
+  assertRemoteControlTargetEnabled(deviceId);
+  const invoke = (): Promise<InvokeResultPayload> => {
+    if (!client) throw new Error('[DEVICE_LINK_NOT_CONNECTED] device-link client not initialized');
+    return client.invoke(deviceId, { channel, args }, INVOKE_TIMEOUT_OVERRIDES_MS[channel]);
+  };
+  return invokeWithClosedLinkRecovery(
+    invoke,
+    () => openRemoteLink(deviceId),
+    () => assertRemoteControlTargetEnabled(deviceId),
+  );
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply apps/desktop/src/main/device-link/index.ts:953]: 已在 `bae490432` 修复：`invokeWithClosedLinkRecovery` 在 reopen 完成后执行 `beforeRetry`；若撤权守卫失败，会 best-effort 调用 `closeRemoteLink` 清理已建立的临时链路，保留原始 `DEVICE_LINK_CONTROL_DISABLED`，且不发送第二次业务 invoke。新增回归测试覆盖该竞态。

File: apps/desktop/src/main/device-link/index.ts [hunk 2]
```diff
@@ -931,6 +951,12 @@ export async function remoteSubscribe(
   topics: string[],
 ): Promise<InvokeResultPayload> {
   assertNotStandby();
+  assertRemoteControlTargetEnabled(deviceId);
+  if (!client) throw new Error('[DEVICE_LINK_NOT_CONNECTED] device-link client not initialized');
+  if (requiresSessionLink(topics) && !client.isLinkReady(deviceId)) {
+    await openRemoteLink(deviceId);
```
[the remark this answers is not in the document]
Comment (dashhuang) [reply apps/desktop/src/main/device-link/index.ts:968]: 已在 `bae490432` 修复：`requiresSessionLink` 改用协议层 `parseFsWatchTopic`，有效的 `fs-watch:<workdir>` 在 link 未 ready 时会先建立 streaming link；空的 `fs-watch:` 不触发建链。新增测试覆盖这两个边界。

**Review by MagicLizi — APPROVED** [review]
Code review passed — well-structured retry-once recovery for LINK_NOT_OPEN with proper inFlight guard, bounded legacy response allowlist, and comprehensive unit tests covering all edge cases. No issues found.

**Comment by MagicLizi** [comment]
断链重连的这种 retry-once + fail-closed 思路很干净——既不冒重复写入的风险，又让控制端体验变成了无感恢复。测试覆盖到 inFlight 边界和撤权竞态也省了后续排查的麻烦 👍
